components/suggest/src/store.rs

```diff
@@ -0,0 +1,421 @@
+        if query.keyword.is_empty() {
+            return Ok(Vec::new());
+        }
+        Ok(self
+            .dbs()?
+            .reader
+            .fetch_by_keyword(&query.keyword)?
+            .into_iter()
+            .filter(|suggestion| {
+                (suggestion.is_sponsored && query.include_sponsored)
+                    || (!suggestion.is_sponsored && query.include_non_sponsored)
+            })
+            .collect())
+    }
+
+    /// Interrupts any ongoing queries. This should be called when the
+    /// user types a new keyword into the address bar, to ensure that they
+    /// see fresh suggestions as they type.
+    pub fn interrupt(&self) {
+        if let Some(dbs) = self.dbs.get() {
+            // Only interrupt if the databases are already open.
+            dbs.reader.interrupt_handle.interrupt();
+        }
+    }
+
+    /// Ingests new suggestions from Remote Settings. `limits` can be used to
+    /// constrain the amount of work done.
+    pub fn ingest(&self, limits: &IngestLimits) -> SuggestApiResult<()> {
+        Ok(self.ingest_inner(limits)?)
+    }
+
+    fn ingest_inner(&self, limits: &IngestLimits) -> Result<()> {
+        let writer = &self.dbs()?.writer;
+        let scope = writer.interrupt_handle.begin_interrupt_scope()?;
+
+        let mut options = GetItemsOptions::new();
+        // Remote Settings returns records in descending modification order
+        // (newest first), but we want them in ascending order (oldest first),
+        // so that we can eventually resume fetching where we left off.
+        options.sort("last_modified", SortOrder::Ascending);
+        if let Some(last_fetch) = writer.get_meta::<u64>(LAST_FETCH_META_KEY)? {
+            // Only fetch changes since our last fetch. If our last fetch was
+            // interrupted, we'll pick up where we left off.
+            options.gt("last_modified", last_fetch.to_string());
+        }
+        if let Some(records) = &limits.records {
+            options.limit(*records);
+        }
+
+        scope.err_if_interrupted()?;
+        let records = self
+            .settings_client
+            .get_records_raw_with_options(&options)?
+            .json::<SuggestRecordsResponse>()?
+            .data;
+        for record in &records {
+            scope.err_if_interrupted()?;
+            match record {
+                FetchedChange::Record(SuggestRecord::Data {
+                    id: record_id,
+                    last_modified,
+                    attachment,
+                }) => {
+                    // Drop any suggestions that we previously ingested from
+                    // this record's attachment. Suggestions don't have a
+                    // stable identifier, and determining which suggestions in
+                    // the attachment actually changed is more complicated than
+                    // dropping and re-ingesting all of them.
+                    writer.drop_suggestions(record_id)?;
```

Comment:
to follow up on my comment above, not being able to do `.drop_suggestions()` in a transaction seems a shame from a perf POV?

Also, you hold the writer mutex while doing http, which is probably OK from a correctness POV, but depending on the foreign bindings implementations, does given the lock might be held for many seconds, does it make sense to "queue" multiple calls to ingest, or might it be worth doing `try_lock()` and failing if an operation is in progress?

Reply:
> Also, you hold the writer mutex while doing http,

Doh - of course you don't - sorry about that :) But the point remains. I do like the db abstractions here!